---
title: "PENDA: PErsoNalized Data Analysis"
subtitle: "Performing personalized data analysis with `penda`"
author: "Magali Richard, ClÃ©mentine Decamps, Florent Chuffart, Daniel Jost"
contact: 
date: "`r Sys.Date()`"
output: 
  rmarkdown::html_document:
    toc: true
    toc_float: true
    toc_depth: 3
    number_sections: true
vignette: >
  %\VignetteEngine{knitr::rmarkdown}
  \usepackage[utf8]{inputenc}
---

```{r echo=FALSE, eval=TRUE}
knitr::opts_chunk$set(collapse=TRUE, comment = "#>", fig.width=9, fig.height=6, eval=TRUE, echo=FALSE, results="hide", dpi=75)
layout(1, respect=TRUE)
penda::draw_penda()
```


# Introduction

`penda` (**PE**rso**N**alized **D**ifferential **A**nalysis ) is an open-access R package that detects gene deregulation in individual samples compared to a set of reference, control samples. This tutorial aims at providing to non-expert users basic informations and illustrations on how to run the package.

How to cite: Richard et al. (2019) PenDa, a rank-based method for Personalized Differential Analysis: application to lung cancer, in submission.

# Dataset

The dataset used to illustrated the method corresponds to the transcriptomes of 3000 genes (RNAseq counts, normalized with DESeq2) for 40 normal, control samples and 40 tumorous samples taken from the TCGA study of lung adenocarcinoma [PMID:25079552].

`data_ctrl` is a data matrix containing the normalized counts of each control sample. 
The rownames of the matrix correspond to the gene_symbol, the colnames indicate the sample ID.

```{r echo=TRUE}
data_ctrl = penda::penda_data_ctrl
```

```{r results="verbatim"}
if (exists("nb_genes")) {
  set.seed(1)
  idx_genes = sample(rownames(data_ctrl), nb_genes)
} else {
  idx_genes = rownames(data_ctrl)
}
# idx_ctrl = sample(colnames(data_ctrl), 20)
idx_ctrl = colnames(data_ctrl)
data_ctrl = data_ctrl[idx_genes,idx_ctrl]

head(data_ctrl[,1:3])
dim(data_ctrl)
```

`data_case` is a data matrix containing the normalized counts of each tumor sample. 
The rownames of the matrix correspond to the gene_symbol, the colnames indicate the sample ID.

```{r echo=TRUE}
data_case = penda::penda_data_case
```

```{r results="verbatim"}
#idx_case = sample(colnames(data_case), 30)
# idx_case = colnames(data_case)
data_case = data_case[rownames(data_ctrl),]
head(data_case[,1:3])
dim(data_case)
```


# Method

`penda` performs a 3-steps analysis: 

1. Data filtering and creation of the dataset

2. Relative gene ordering

3. Differential expression testing


## Data filtering

```{r label="penda::make_dataset"}
threshold_dataset = 0.99
if (!exists("Penda_dataset")) Penda_dataset = penda::make_dataset(data_ctrl, data_case, detectlowvalue = TRUE, detectNA = TRUE, threshold = threshold_dataset)
data_ctrl_f = Penda_dataset$data_ctrl
data_case_f = Penda_dataset$data_case

dim(data_ctrl_f)
dim(data_case_f)
```

The function `make_dataset` contains three steps to prepare the data for the analysis.  

  - `detect_na_value` removes rows and columns (ie, genes and samples) of the data matrices that contain more than `r "threshold"`% (default value = `r threshold_dataset`) of NA (Not Available) value. <br/>
  - `detect_zero_value` removes genes with very low expression in the majority of samples (controls and cases), ie. genes whose expression is lower than `r "val_min"` in `r "threshold"`% of all the samples. By default it uses the function `normalmixEM` to estimate the value of `r "val_min"` using all the log2-transformed count data but this parameter can also be tuned manually by the user.
  - `rank_genes` sorts the genes based on the median of gene expression in controls. This step is essential for the proper functioning of `penda`.


## Relative gene ordering

```{r, label="penda::compute_lower_and_higher_lists"}
threshold_LH = 0.99
s_max = 30
if (!exists("L_H_list")) L_H_list = penda::compute_lower_and_higher_lists(data_ctrl_f, threshold = threshold_LH, s_max = s_max)
```

The function `compute_lower_and_higher_lists` computes two matrices of lists L and H based on the filtered control dataset (`data_ctrl_f`).  

Each row of the **L** matrix contains a list of at most `s_max` (default value = `r s_max`) genes (characterized by their ids) whose expressions are **lower** than that of the gene associated to the corresponding row, in at least `threshold_LH` (default value = `r threshold_LH*100` %) of the control samples.<br/>
Each row of the **H** matrix contains a list of at most `s_max` (default value = `r s_max`) genes (characterized by their ids) whose expressions are **higher** than that of the gene associated to the corresponding row, in at least `threshold_LH` (default value = `r threshold_LH*100` %) of the control samples.<br/>

Bellow are the histograms of the size of L and H lists fo all the genes:


```{r, results="verbatim"}
layout(matrix(1:2, 1), respect=TRUE)
hist(rowSums(L_H_list$L != 0), xlab = "nb of L genes")
hist(rowSums(L_H_list$H != 0), xlab = "nb of H genes")
```


## Differential expression testing

```{r, label="penda::penda_test", echo = FALSE, results = "hide"}
controls =  data_ctrl_f
samples = data_case_f
threshold = 0.5
iterations =  20 
L_H_list =  L_H_list 
quant_test =  0.05
factor_test = 1.2

penda_res = penda::penda_test(samples = samples, 
                  controls = controls,
                   threshold = threshold, 
                   iterations =  iterations, 
                   L_H_list =  L_H_list, 
                   quant_test =  quant_test,
                   factor_test = factor_test)
```

The function  `penda_test` infers for each gene and for each sample of the `data_case` matrix its deregulation status (up-regulation, down-regulation or no deregulation). This function analyses case samples one by one. It is based on the `L_H_list` and tracks for changes in relative ordering in the sample of interest. If these changes exceed the given `threshold`, the gene of interest is considered as deregulated. 

By default, the `threshold` parameter is set to `r threshold` but we strongly advise users to use the vignette `vignette simulation` to adjust this parameter to the user-specific data.

Results are in the form of two matrices `$down_genes` and `$up_genes`. Each row corresponds to a gene and each column to a case sample. A TRUE entry in these matrices meands that the corresponding genes is deregulated (down or up-regulated) in the corresponding samples. 





```{r, echo = FALSE, warning = FALSE}
generate_data_bypatient = function(D_list, U_list){
  down = colSums(D_list)
  up = colSums(U_list)
  total = down + up
  patient_names = colnames(D_list)
  patients = rep(factor(patient_names , levels = patient_names [order(total)]), 3)
  variable = c(rep("down", length(patient_names)),
               rep("up", length(patient_names)),
               rep("total", length(patient_names)))
  value = c(down, up, total)
  return(data.frame(patients = patients,
                    variable = variable,
                    value = value))
}

data_bypatient = generate_data_bypatient(D_list = penda_res$down_genes,
                                         U_list = penda_res$up_genes)

plot_figure = function(data_patients){
library(ggplot2)
mytheme <- theme(panel.background = element_blank(),
  panel.grid.major = element_line(colour="black", size = (0.1)),
                 panel.grid.minor = element_blank())

p1 = ggplot(data_patients, aes(x = patients, y = value)) + 
    geom_line(aes(group = variable), colour = "grey80") + # theme_minimal(base_size = 20) +
    mytheme + ylim(0, 10000) + ylab("Number of  genes") + xlab("Patients") +
theme(axis.text.x = element_blank()) + geom_point(aes(colour = variable), 
    size = 0.5) + ggtitle("Degree of deregulation per patient") +  
scale_y_continuous(breaks = seq(0, 10000, 2000)) + scale_x_discrete(breaks = NULL) +
   scale_colour_manual(name  ="Gene deregulation", 
                      values = c("green", "darkblue", "red"), 
                       labels=c("DOWN", "UP & DOWN", "UP")) 
 return(p1)
}

plot_figure(data_patients = data_bypatient)

```

```{r fig.width=9, fig.height=9, echo = FALSE, warning = FALSE}

plot_heatmap_hclust = function (data) {
  sum(apply(is.na(data), 1, any))
  data = data[!apply(is.na(data), 1, any), ]

  # clustering base on correlation for tissues
  tmp_d = data
  tmp_d = t(tmp_d) - apply(tmp_d, 2, mean)
  tmp_d = t(tmp_d)
  tmp_d = cor(tmp_d, method="pe")
  dim(tmp_d)
  hc_col = hclust(dist(1 - tmp_d), method="complete")

  Colv = as.dendrogram(hc_col)
  dendrogram="col"      

  # clustering base on eucl. dist. for genes
  d = dist(data)
  hc_row = hclust(d, method="complete")
  Rowv = as.dendrogram(hc_row)
  dendrogram="both"      

  # col
  colors=c("blue", "gray", "red")
  cols = colorRampPalette(colors)(20)

  foo = gplots::heatmap.2(data, Rowv=Rowv, Colv=Colv, dendrogram="col", trace="none", col=cols,
                          labRow = FALSE,labCol = FALSE,
                          main=paste0("Penda (", nrow(data), " genes x ", ncol(data), " samples)"), mar=c(10,5), useRaster=TRUE)
}


plot_heatmap_hclust(data = penda_res$down_genes - penda_res$up_genes)
```


# Material and methods

*This paragraph is automatically generated by the vignette to specify the method and data filtering parameters. It can be directly cut and paste to the "material and methods" section of the user analysis.*

The Penda vignette of the penda package version 1.0 was executed on `r floor(Penda_dataset$info["init_nb_genes"])`  genes, using `r floor(Penda_dataset$info["init_nb_ctrls"])` control samples and `r floor(Penda_dataset$info["init_nb_cases"])` case samples.

The data set was pretreated as following: 
`r floor(Penda_dataset$info["nb_genes_NA"])` genes and 
`r floor(Penda_dataset$info["nb_patients_NA"])` patients was removed for NA values, 
and `r floor(Penda_dataset$info["nb_genes_0"])` genes was removed for low value 
(under the threshold `val_min` of `r Penda_dataset$info["val_min"]`) in at least `r Penda_dataset$info["threshold"]*100` % of cases.

`r ncol(controls)` controls were used to generate L and H lists using the following parameters: threshold LH = `r threshold_LH` and s_max = `r s_max`.

The penda method was then applied on `r ncol(data_case)` cases. With the following set of parameters: quantile = `r quant_test`, factor = `r factor_test` and threshold = `r threshold`.




# Session Information

```{r, results="verbatim"}
sessionInfo()
```



