---
title: "Advanced User - Performing PErsoNalized Data Analysis with `penda`"
author: "Magali Richard, Florent Chuffart, Clémentine Decamps, Daniel Jost"
date: "`r Sys.Date()`"
output: 
  rmarkdown::html_document:
    toc: true
    toc_float: true
    toc_depth: 3
    number_sections: true
---

```{r, echo=FALSE}
draw_penda = function(){
  library(ggplot2)
  tetey = c(4, 5, 5.8, 5.8, 5.4, 5, 4, 3, 1.5, 1.3, 1.5, 2.5, 4)
  tetex = c(1, 1.75, 2.9, 4, 4.8, 5.2, 5.8, 6, 5, 3.8, 2.3, 1.2, 1)
  orgy = c(4.5, 5.2, 6, 6.2, 5.8, 5, 4.5)
  orgx = c(1.35, 1, 1.2, 2.3, 2.9, 1.75, 1.35)
  ordy = c(5.8, 6.2, 6, 5.6, 4.8, 4.2, 5, 5.4, 5.8)
  ordx = c(4, 4.8, 5.3, 5.9, 6, 5.7, 5.2, 4.8, 4)

  ydx = c(3.9, 4.2, 4.8, 5, 4.8, 4.2, 3.8, 3.9)
  ydy = c(4, 4.4, 4.2, 4, 3.3, 3.2, 3.5, 4)

  ygx = c(2.5, 2.2, 2.5, 3.2, 3.5, 3.2, 2.5)
  ygy = c(3.2, 4, 4.3, 4.3, 4, 3.3, 3.2 )

  points = data.frame(x = c(2.8, 3.5, 4.3)
                    ,y = c(3.8, 2.8, 3.8))

  graph = ggplot() + geom_path(aes(tetex, tetey)) + geom_path(aes(orgx, orgy, size = 3)) + geom_path(aes(ordx, ordy, size = 3)) + geom_path(aes(ydx, ydy, size = 3)) + geom_path(aes(ygx, ygy, size = 3)) + geom_point(data = points, aes(x = x, y = y, size = 3)) +
    theme_bw() +
    theme(legend.position="none") +
    theme(axis.title.x=element_blank(),
        axis.text.x=element_blank(),
        axis.ticks.x=element_blank(),
        axis.title.y=element_blank(),
        axis.text.y=element_blank(),
        axis.ticks.y=element_blank()) +
     ggtitle("penda",subtitle = "PErsoNalized Data Analysis" ) +
  theme(plot.title = element_text(lineheight=10, face="bold", hjust = 0.5), plot.subtitle = element_text(lineheight=.8, face="italic", hjust = 0.5)) 
  
  return(graph)
}
draw_penda()
```

# Introduction

`penda` (**PE**rso**N**alized **D**ifferential **A**nalysis ) is an open-access analysis software (R package). The purpose of this tutorial is to explain how choose different parameters according to your data for an advanced user.  

<span style="color:red">Mag_com: Je pense qu'il faut expliquer un peu plus ici à quoi sert la vignette puis diviser le plan en 2 : (i) generate your simulated dataset avec (ia) simple simulation et (ib) simulation inspired by real dataset puis (ii) define optimal parameters for your dataset. Pour ecrire ce paragraphe, imagine que tu es un nouvel utilisateur qui découvre la vignette. </span>

# Dataset

The dataset used for this example correspond to the transcriptomes (normalized count, using DESeq2 method) of 44 patients with lung adenocarcinoma (PMID:25079552), including control and pathological samples. 

`data_ctrl` is a data matrix containing normalized counts for each control sample. 
The rownames of the matrix correspond the gene_symbol , the colnames indicate the sample ID.

```{r, results="verbatim"}
data_ctrl = penda::data_ctrl
head(data_ctrl[,1:5])
is.matrix(data_ctrl)
dim(data_ctrl)
```

`data_case` is a data matrix containing normalized counts for each tumor sample. 
The rownames of the matrix correspond the gene_symbol , the colnames indicate the sample ID.

```{r, results="verbatim"}
data_case = penda::data_case
head(data_case[,1:5])
is.matrix(data_case)
dim(data_case)
```

You can use the function `detect_zero_value` to remove genes with very low expression, or the function `detect_na_value` to remove those with high rate of NA.

We need control samples to do simulations and be able to compute the error rate. Here, we extract the five last patients of data_ctrl in `data_simu`.

```{r, tidy=TRUE, tidy.opts=list(blank=FALSE, width.cutoff=55)}
data_simu = data_ctrl[, (ncol(data_ctrl)-4):ncol(data_ctrl)]
colnames(data_simu) = c("P1","P2","P3","P4","P5")

data_ctrl = data_ctrl[, -((ncol(data_ctrl)-4):ncol(data_ctrl))]
```


#Simulation of dysregulation

To compute the real error rate, we need to use simulations. 
The function `complex_simulation` uses the real distribution of difference between data_case and data_ctrl to simulate the proportion and the value of the dysregulation. 

The simulation returns the vector of initial data, the vector of datas with modifications and the index of modified data. 

```{r, tidy=TRUE, tidy.opts=list(blank=FALSE, width.cutoff=55)}

if (!file.exists("tmp/simulation.rds")){
  simulation = penda::complex_simulation(data_ctrl, data_case, data_simu)
  saveRDS(simulation , "tmp/simulation.rds")
}

simulation = readRDS("tmp/simulation.rds")

```

#Gene ranking in control

<span style="color:red">Mag_com: Il faut etre plus clair sur les parametres qu'on peut faire varier et l'impact de ces variations </span>


Now, we have to rank gene expression in normal tissue to use the Penda method.

We used the function `compute_down_and_up_list` which a list of two matrices. 

The matrix **D** corresponds to the list (i.e true) of downregulated genes for each considered genes (columns).
The matrix **U** corresponds to the list (i.e true) of upregulated genes for each considered genes (columns).

For each, the list of **D** genes (resp. **U** genes) corresponds to the closest genes (up to `s_max` genes) that are down (resp. up) regulated in at least `threshold` % of the control samples. 

```{r, tidy=TRUE, tidy.opts=list(blank=FALSE, width.cutoff=55)}

if (!file.exists("tmp/D_U_list_simu.rds")){
  D_U_list = penda::compute_down_and_up_list(data_ctrl, threshold = 0.99, s_max = 50)
  saveRDS(D_U_list, "tmp/D_U_list_simu.rds")
}
D_U_list = readRDS("tmp/D_U_list_simu.rds")

```


#Parameters of the quantile method

<span style="color:red">Mag_com: peut etre qu'on pourrait inclure une figure ici pour un schéma, avec tous les parametres... </span>

First parameters to test are those of the quantile method. This method is used in the Penda method when we don't have **D** or **U** list. There are two parameters, the `quantile` of the gene expression distribution in the control, and the `factor` which modulate this quantile value.

To vary these parameters, we use the function `choose_quantile` which, for one patient, vary the quantile value for a fixed factor. This function compute then, for each quantile, three error rates.
The **FDR**, False Discovery Rate, the proportion of false positive in all the positive results.
The **TPR**, True Positive Rate, the proportion of true positives detected in results.
The **FPR**, False Positive Rate, the proportion of false positive in all the negative results. 

With these values, we can make a curve and choose the better quantile and the better factor to fit the experimental requirements (low FDR and high TPR). <span style="color:red">Mag_com: ce n'est pas très bien dit </span>

<span style="color:red">Mag_com: pour le chunk suivant, tu peux également écrire un vecteur de facteur, et enrober tout cela dans une fonction qui prend en argument : quantile_values, factor et FDR_goal. Attention, il y a un probleme dans la legende de la figure. Tu peux egalement faire apparaitre de maniere claire sur la figure les parametres que tu choisis finalement. </span>

```{r, tidy=TRUE, tidy.opts=list(blank=FALSE, width.cutoff=55)}

quantile_values = c(0, 0.01, 0.02, 0.03, 0.05, 0.06, 0.08, 0.1, 0.13, 0.15, 0.2, 0.25, 0.30, 0.35, 0.4, 0.45)

simup1 = list(initial_data = simulation$initial_data[,1], simulated_data = simulation$simulated_data[,1])

qf1 = penda::choose_quantile(data_ctrl, simup1, FDR_goal = 0.06, factor = 1, quantile_values = quantile_values)
qf15 = penda::choose_quantile(data_ctrl, simup1, FDR_goal = 0.06, factor = 1.5, quantile_values = quantile_values)
qf2 = penda::choose_quantile(data_ctrl, simup1, FDR_goal = 0.06, factor = 2, quantile_values = quantile_values)
qf25 = penda::choose_quantile(data_ctrl, simup1, FDR_goal = 0.06, factor = 2.5, quantile_values = quantile_values)
qf3 = penda::choose_quantile(data_ctrl, simup1, FDR_goal = 0.06, factor = 3, quantile_values = quantile_values)

df_quant = data.frame(quantile = factor(c(qf1[,1], qf15[,1], qf2[,1], qf25[,1], qf3[,1]))
                      , FDR = as.numeric(c(qf1[,2], qf15[,2], qf2[,2], qf25[,2], qf3[,2]))
                      , TPR = as.numeric(c(qf1[,3], qf15[,3], qf2[,3], qf25[,3], qf3[,3]))
                      , facteur = factor(c(rep(c(1,1.5, 2, 2.5, 3), each = 16)))
)

library(ggplot2)
ggplot(df_quant, aes(x = TPR, y = FDR, color = quantile, group = facteur, shape = facteur)) + geom_point() + geom_line()


```

#Threshold of the Penda test

<span style="color:red">Mag_com: on peut aussi mettre une figure ici. </span>

The last important parameter to choose is the `threshold` of the test. This threshold is the importance of the change of rank necessary to assert the deregulation of a gene. 
The function `choose_threshold` make the Penda test to different threshold values and compute the FDR, TPR and FPR. 

With these values, we can make a curve and choose the better parameter to fit the experimental requirements (low FDR and high TPR).

<span style="color:red">Mag_com: Le code est un peu trop dense pour une vignette, voit si tu peux simplifier... Tu peux aussi mettre en evidence sur la figure les parametres choisis. </span>

```{r, tidy=TRUE, tidy.opts=list(blank=FALSE, width.cutoff=55)}

if (!file.exists("tmp/which_threshold.rds")){
  threshold_values = c(0, 0.05, 0.1, 0.3, 0.5, 0.6, 0.7, 0.8, 0.9)
  simup1 = list(initial_data = simulation$initial_data[,1], simulated_data = simulation$simulated_data[,1])
  which_threshold = penda::choose_threshold(data_ctrl, D_U_list, 30, simup1, threshold_values, quant_test = 0.03, factor_test = 1)
  saveRDS(which_threshold, "tmp/which_threshold.rds")
}

which_threshold = readRDS("tmp/which_threshold.rds")

library(ggplot2)
df = data.frame(patient = which_threshold[,1],
                threshold = which_threshold[,2],
                FDR = as.numeric(which_threshold[,3]),
                TPR = as.numeric(which_threshold[,4]))


ggplot(df, aes(x = TPR, y = FDR, color = threshold, group = patient)) +
  geom_point() + geom_line()+ theme_minimal()

```

With these paramaters, you can now start your analysis on all the datas. 
